feat: Add `set_as_inventory_for` api function

Lazerbeak12345 opened the pull request:
Adds the `Form:set_as_inventory_for(player, [ctx])` api.

Also includes

- `Form:unset_as_inventory_for(player)`
- Extends private `show_form` to account for inv edge cases
- ~~Extends `Form:update` to account for inv edge cases~~
- ~~Extends `Form:update_where` to account for inv edge cases~~
- Extends private `on_fs_input` to account for inv edge cases
- Adds abstraction functions
- Add feature to `on_leaveplayer` to delete the inv formspec as well
- Add a few unit tests for related features

see #1

Lazerbeak12345 committed:
feat: Add `set_as_inventory_for` api function

Adds the `Form:set_as_inventory_for(player, [ctx])` api.

Also includes

- `Form:unset_as_inventory_for(player)`
- Extends private `show_form` to account for inv edge cases
- Extends `Form:update` to account for inv edge cases
- Extends `Form:update_where` to account for inv edge cases
- Extends private `on_fs_input` to account for inv edge cases
- Adds several abstraction functions
- Add feature to `on_leaveplayer` to delete the inv formspec as well
- Add a few unit tests for related features

see #1

Lazerbeak12345 commented:
Doesn't include documentation right now

Lazerbeak12345 commented on init.lua:
```diff
@@ -799,6 +820,11 @@ function Form:close_hud(player)
 
+function Form:unset_as_inventory_for(player)
+    open_inv_formspecs[player:get_player_name()] = nil
+    player:set_inventory_formspec("") -- TODO should this restore what it was before? This disables inventory as-is until set by something else.
```
I'm not sure what you'd like this to do here.

luk3yx replied:
I think this is okay, I'd be happy with it not changing the inventory formspec at all here but maybe it's a good idea to set it to `""` prevent confusion.

Maybe this function should do a check like `form:close` does so that you can't call unset_as_inventory_for with the wrong form object?

Lazerbeak12345 replied:
Here's a test of what that would look like:

```lua
function Form:unset_as_inventory_for(player)
    local name = player:get_player_name()
    local form_info = open_inv_formspecs[name]
    if form_info and form_info.self == self then
        open_inv_formspecs[name] = nil
        player:set_inventory_formspec("")
    end
end
```

My concern is that it feels like since it's more or less the same as `Form:close`. Perhaps it'd be better to merge them? Or the duplication doesn't matter. Idk.

luk3yx replied:
The duplication here is okay, I think that trying to remove it would just make things more complicated.

Lazerbeak12345 replied:
resolved in 0789c66

luk3yx commented on init.lua:
```diff
@@ -811,18 +837,21 @@ local function update_form(self, player, form_info)
 end
 
 function Form:update(player)
-    local form_info = open_formspecs[player:get_player_name()]
+    local player_name = player:get_player_name()
+    local form_info = open_formspecs[player_name] or open_inv_formspecs[player_name]
     if form_info and form_info.self == self then
         update_form(self, player, form_info)
     end
 end
 
 function Form:update_where(func)
-    for name, form_info in pairs(open_formspecs) do
-        if form_info.self == self then
-            local player = minetest.get_player_by_name(name)
-            if player and func(player, form_info.ctx) then
-                update_form(self, player, form_info)
+    for _ , open_formspec_table in ipairs({open_formspecs, open_inv_formspecs}) do
+        for name, form_info in pairs(open_formspec_table) do
+            if form_info.self == self then
+                local player = minetest.get_player_by_name(name)
+                if player and func(player, form_info.ctx) then
+                    update_form(self, player, form_info)
+                end
```
I don't think it's worth making `form:update` and `form:update_for` work with inventory formspecs, they don't work with HUDs at the moment and it should be possible to just call `set_as_inventory_for` again.

Lazerbeak12345 replied:
Good point. There's actually a bug I missed here where it will call `minetest.show_formspec` on the updated form. Not sure if that's how it's supposed to be done.

Lazerbeak12345 replied:
fixed

luk3yx commented on init.lua:
```diff
@@ -878,6 +918,7 @@ end
 local function on_leaveplayer(player)
     open_formspecs[player:get_player_name()] = nil
+    open_inv_formspecs[player:get_player_name()] = nil
 end
```
```suggestion
local function on_leaveplayer(player)
    local name = player:get_player_name()
    open_formspecs[name] = nil
    open_inv_formspecs[name] = nil
end
```

luk3yx commented on init.lua:
```diff
@@ -835,7 +864,14 @@ end
+    else
+        form_info = open_inv_formspecs[name]
+        inv = true
+    end
```
```suggestion
    local form_infos = formname == "" and open_inv_formspecs or open_formspecs
    local form_info = form_infos[name]
```

luk3yx commented on init.lua:
```diff
@@ -866,9 +902,13 @@ local function on_fs_input(player, formname, fields)
+        if open_inv_formspecs[name] ~= form_info then return true end
+    else
+        if open_formspecs[name] ~= form_info then return true end
+    end
```
```suggestion
    if form_infos[name] ~= form_info then return true end
```

luk3yx commented on init.lua:
```diff
@@ -786,6 +803,10 @@ function Form:show_hud(player, ctx)
 
+function Form:set_as_inventory_for(player, ctx)
+    set_inv_form(self, player, ctx or {})
+end
```
I wonder whether it would be better to merge `set_inv_form` into this function (since it isn't used anywhere else) and make this function preserve the previous `ctx` (if the form is already set as the inventory and if the `ctx` argument isn't present) and `auto_name_id` (even if a different form was previously set).

Lazerbeak12345 replied:
done

luk3yx commented on init.lua:
```diff
@@ -866,9 +902,13 @@ local function on_fs_input(player, formname, fields)
+    end
 
-    if fields.quit then
+    if fields.quit and not inv then
```
```suggestion
    if formname == "" then
        -- Special case for inventory forms
        if redraw_fs then
            form_info.self:set_as_inventory_for(player)
        end
    elseif fields.quit then
```

luk3yx commented on init.lua:
```diff
@@ -757,13 +756,31 @@ local function show_form(self, player, formname, ctx, auto_name_id)
     -- if DEBUG_MODE then
     --     print(t3 - t, t2 - t, t3 - t2)
     -- end
+    return  fs, form_info
```
```suggestion
    return fs, form_info
```

luk3yx reviewed (commented):
Thank you for your PR! I've made some suggestions to make the code slightly more compact and to hopefully fix a bug (`update_form` will call `show_formspec` and not change the inventory formspec).

Lazerbeak12345 committed:
fix(on_leaveplayer): only get player's name once while they leave

Co-authored-by: luk3yx <luk3yx@users.noreply.github.com>

Lazerbeak12345 committed:
fix: update bugs and some restructuring

Applies suggestions from code review

Co-authored-by: luk3yx <luk3yx@users.noreply.github.com>

Lazerbeak12345 committed:
fix: other bugs and other refactor

Lazerbeak12345 committed:
test(update): Don't assume update does anything

Lazerbeak12345 commented:
Updated readme to reflect changes in `update` support.

luk3yx commented on init.lua:
```diff
@@ -786,6 +793,16 @@ function Form:show_hud(player, ctx)
+local open_inv_formspecs = {}
+function Form:set_as_inventory_for(player, ctx)
+    local name = player:get_player_name()
+    -- Formname of "" is inventory
+    local fs, form_info = prepare_form(self, player, "", ctx or {})
```
Maybe this function should preserve `ctx` if the same form is already set as the inventory?

```suggestion
    local old_form_info = open_inv_formspecs[name]
    if not ctx and old_form_info and old_form_info.self == self then
        ctx = old_form_info.ctx
    end

    -- Formname of "" is inventory
    local fs, form_info = prepare_form(self, player, "", ctx or {},
        old_form_info and old_form_info.auto_name_id)
```

`auto_name_id` stores the next number to use for automatic button naming to work around race conditions that would happen if button names were reused.

I may move the `auto_name_id < 1e6` check from `update_form` into `_render` later so it doesn't have to be duplicated here.

Lazerbeak12345 committed:
fix: Ensure CTX is retained when updating inventory

Co-authored-by: luk3yx <luk3yx@users.noreply.github.com>

Lazerbeak12345 committed:
feat(unset_as_inventory_for): ensure correct form

With this fix you can't call `Form:unset_as_inventory_for` with the
wrong form object.

Co-authored-by: luk3yx <luk3yx@users.noreply.github.com>

luk3yx committed:
Update Form:update()

luk3yx committed:
Make luacheck happy

Maybe I should increase the line length limit one day

luk3yx reviewed (approved):
I'll try and test this PR and merge it soon

luk3yx committed:
Make auto_name_id wrap around in _render

Lazerbeak12345 commented:
Would you like me to add the documentation to the readme?

luk3yx commented:
> Would you like me to add the documentation to the readme?

Sure, in the "experimental features" section.

luk3yx committed:
Bugfix

If you move a scrollbar or check a checkbox and then close the inventory (without doing anything that would redraw the form), the position of scrollbars and states of checkboxes won't be remembered the next time you open the inventory.

This commit makes flow redraw inventory formspecs when the player closes them if anything in ctx.form has been changed by flow since it was last redrawn.

luk3yx commented:
I've tested this PR and it works okay, I fixed a bug with remembering the values of scrollbars and checkboxes when the inventory is closed.

Lazerbeak12345 committed:
docs(inventory): document new inventory formspec

luk3yx commented on README.md:
```diff
@@ -399,3 +399,34 @@ gui.Stack{
+### Set form as player inventory
+
+A form can be set as the player inventory. Flow internally generates the
+formspec and passes it to `player:set_inventory_formspec()`.
```
Perhaps the incompatibility with other inventory mods (like sfinv) should be mentioned?

Lazerbeak12345 replied:
Sure. Something like "This will override your inventory mod?"

luk3yx replied:
Maybe "This will completely replace your inventory and isn't compatible with inventory mods like sfinv."

Lazerbeak12345 replied:
Oh, right. That's even better.

luk3yx commented on README.md:
````diff
@@ -399,3 +399,34 @@ gui.Stack{
+
+```lua
+example_inventory:set_as_inventory_for(player, new_context)
+```
````
This should mention that calling `example_inventory:set_as_inventory_for(player)` will reuse the existing context

Lazerbeak12345 committed:
docs(set inventory): Clarify with more info

Lazerbeak12345 committed:
docs(set inventory): note potential sfinv breakage

luk3yx committed:
Merge branch 'main' into 1-form-set-as-inv-for-player

Lazerbeak12345 commented:
nice. I was just going to do the same and rebase or something.

luk3yx committed:
Update README.md

luk3yx commented:
Thanks again for your PR, good luck with your inventory mod!

Lazerbeak12345 commented:
Yeah, thanks for being so responsive! I'll watch for a new release on contentdb, and I'll likely stick around for #3 as well, even though at this point the API is enough for my needs.
